akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -367,11 +427,83 @@ object ByteString {
 
     def decodeString(charset: String): String = compact.decodeString(charset)
 
+    def decodeString(charset: Charset): String =
+      compact.decodeString(charset)
+
     private[akka] def writeToOutputStream(os: ObjectOutputStream): Unit = {
       os.writeInt(bytestrings.length)
       bytestrings.foreach(_.writeToOutputStream(os))
     }
 
+    override def take(n: Int): ByteString = {
+      @tailrec def take0(n: Int, b: ByteStringBuilder, bs: Vector[ByteString1]): ByteString =
+        if (bs.isEmpty || n <= 0) b.result
+        else if (n <= bs.head.length) b.append(bs.head.take(n)).result
+        else take0(n - bs.head.length, b.append(bs.head), bs.tail)
+
+      if (n <= 0) ByteString.empty
+      else if (n >= length) this
+      else take0(n, ByteString.newBuilder, bytestrings)
+    }
+
+    override def dropRight(n: Int): ByteString = {
+      if (n <= 0) this
+      else if (n < bytestrings.last.length) {
+        val last = bytestrings.last
+        new ByteStrings(bytestrings.init :+ last.dropRight1(n), length - n)
+      } else {
+        val last = bytestrings.last
+        val remaining = bytestrings.init
+        if (remaining.isEmpty) ByteString.empty
+        else {
+          val s = new ByteStrings(remaining, length - last.length)
+          val remainingToBeDropped = n - last.length
+          s.dropRight(remainingToBeDropped)
+        }
+      }
+    }
+
+    override def slice(from: Int, until: Int): ByteString =
+      if ((from == 0) && (until == length)) this
+      else if (from > length || until <= from) ByteString.empty
+      else {
+        val l = drop(from)
+        val r = l.dropRight(length - until)
+        r
+      }
+
+    override def drop(n: Int): ByteString =
+      if (n <= 0) this
+      else if (n > length) ByteString.empty
+      else drop0(n)
+
+    private def drop0(n: Int): ByteString = {
+      var continue = true
+      var fullDrops = 0
+      var remainingToDrop = n
+      do {
+        val currentLength = bytestrings(fullDrops).length
+        if (remainingToDrop >= currentLength) {
+          fullDrops += 1
+          remainingToDrop -= currentLength
+        } else continue = false
```

Comment:
can't do that as `remainingToDrop > 0 && remainingToDrop < currentLength` is a value I need to keep.
Perhaps a crazy trick with making it `remainingToDrop = -remainingToDrop` would work hm, since then I still know I need to drop a bit